elliptic_curve/src/point.rs

```diff
@@ -104,130 +141,255 @@ impl Add for Point {
         }
 
         let (x1, y1, x2, y2) = (
-            self.x.clone().unwrap(), 
-            self.y.clone().unwrap(), 
-            rhs.x.clone().unwrap(), 
-            rhs.y.clone().unwrap());
+            &self.x.clone().unwrap(),
+            &self.y.clone().unwrap(),
+            &rhs.x.clone().unwrap(),
+            &rhs.y.clone().unwrap(),
+        );
 
         if self == rhs {
-            let prime = self.a.prime.clone();
+            let prime = &self.a.prime;
             let big3 = FF::new(BigUint::from(3_u32), prime.clone());
             let big2 = FF::new(BigUint::from(2_u32), prime.clone());
-            let doub_y1 = big2 * y1.clone();
-            let sqrt_x1 = x1.clone().pow(2);
-            let a = self.a.clone();
-            let lamda = (big3*sqrt_x1 + a) / (doub_y1); 
-            let x3 = lamda.clone().pow(2) - x1.clone() - x1.clone();    
-            let y3 = lamda.clone() * (x1.clone() - x3.clone()) - y1.clone();
+
+            let doub_y1 = big2 * y1;
+            let sqrt_x1 = x1.pow(2);
+
+            let a = &self.a;
+            let lamda = (big3 * sqrt_x1 + a) / (doub_y1);
+
+            let x3 = lamda.pow(2) - x1 - x1.clone();
+            let y3 = lamda * (x1.clone() - x3.clone()) - y1;
 
             return Self {
                 x: Some(x3),
                 y: Some(y3),
                 a: self.a.clone(),
-                b: self.b.clone()
-            };  
-        } 
+                b: self.b.clone(),
+            };
+        }
 
         let t1 = y2.clone() - y1.clone();
         let t2 = x2.clone() - x1.clone();
+
         let lamda = t1 / t2;
-        let x3 = lamda.clone().pow(2) - x1.clone() - x2.clone();    
+        let x3 = lamda.pow(2) - x1 - x2;
         let y3 = lamda.clone() * (x1.clone() - x3.clone()) - y1.clone();
-        return Self {
+        Self {
             x: Some(x3),
             y: Some(y3),
             a: self.a.clone(),
-            b: self.b.clone()
-            };
+            b: self.b.clone(),
         }
+    }
 }
 
+impl<'a> Add<&'a Self> for ECCPoint {
+    type Output = Self;
+
+    fn add(self, rhs: &Self) -> Self::Output {
+        if self.a != rhs.a || self.b != rhs.b {
+            panic!("not in the same curve")
+        }
+
+        if self.x.is_none() && self.y.is_none() {
+            return rhs.clone();
+        }
+
+        if rhs.x.is_none() && rhs.y.is_none() {
+            return self;
+        }
```

Comment:
If you encounter situations like this, consider using match. It also supports if conditions.
See https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards